torch/csrc/api/include/torch/nn/pimpl.h

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <torch/csrc/utils/variadic.h>
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace torch {
+namespace detail {
+/// This class exists  only to do SFINAE on abstract types `T` that are really
+/// `ModuleHolder<ModuleType>`, because there's no good way to say that `T` is a
+/// `ModuleHolder` over some unknown type `ModuleType`. With this, you can do
+/// enable_if_t<is_base_of<ModuleHolderIndicator, T>::value>::type.
+struct ModuleHolderIndicator {};
+
+template <typename T>
+using disable_if_module_holder_t =
+    disable_if_t<std::is_base_of<ModuleHolderIndicator, decay_t<T>>::value>;
+} // namespace detail
+
+namespace nn {
+
+/// A `ModuleHolder` is essentially a wrapper around `std::shared_ptr<M>` where
+/// `M` is an `nn::Module` subclass, with convenient constructors defined for
+/// the kind of constructions we want to allow for our modules.
+template <typename Contained>
+class ModuleHolder : torch::detail::ModuleHolderIndicator {
+ public:
+  using ContainedType = Contained;
+
+  /// Constructs the `ModuleHolder` with an empty contained value.
+  ModuleHolder() = default;
+
+  /// Single argument constructor of the underlying type.
+  template <typename T>
+  explicit ModuleHolder(T&& t)
+      : impl_(std::make_shared<Contained>(std::forward<T>(t))) {}
+
+  /// Multi-argument constructor. This constructor is special in that the
+  /// expectation is that the constructor of the contained type takes an object
+  /// that can be constructed with the given arguments. For our modules, this is
+  /// always the `Options` struct. For this reason, the arguments are forwarded
```

Comment:
I don't think the comments here are clear enough about expected usage. Maybe point to an explicit example.